src/tile.cpp

```diff
@@ -396,94 +394,96 @@ void Tile::onUpdateTileItem(Item* oldItem, const ItemType& oldType, Item* newIte
+	for (const auto& spectator : spectators) {
+		spectator->onUpdateTileItem(getTile(), thingMapPos, oldItem, oldType, newItem, newType);
 	}
 }
 
-void Tile::onRemoveTileItem(const SpectatorVec& spectators, const std::vector<int32_t>& oldStackPosVector, Item* item)
+void Tile::onRemoveTileItem(const SpectatorVec& spectators, const std::vector<int32_t>& oldStackPosVector,
+                            std::shared_ptr<Item> item)
 {
 	if (item->hasProperty(CONST_PROP_MOVEABLE) || item->getContainer()) {
-		auto it = g_game.browseFields.find(this);
+		auto it = g_game.browseFields.find(getTile());
 		if (it != g_game.browseFields.end()) {
 			it->second->removeThing(item, item->getItemCount());
 		}
 	}
 
 	resetTileFlags(item);
 
-	const Position& cylinderMapPos = getPosition();
+	const Position& thingMapPos = getPosition();
 	const ItemType& iType = Item::items[item->getID()];
 
 	// send to client
 	size_t i = 0;
-	for (Creature* spectator : spectators) {
-		if (Player* tmpPlayer = spectator->getPlayer()) {
-			tmpPlayer->sendRemoveTileThing(cylinderMapPos, oldStackPosVector[i++]);
+	for (const auto& spectator : spectators) {
+		if (auto tmpPlayer = spectator->getPlayer()) {
+			tmpPlayer->sendRemoveTileThing(thingMapPos, oldStackPosVector[i++]);
 		}
 	}
 
 	// event methods
-	for (Creature* spectator : spectators) {
-		spectator->onRemoveTileItem(this, cylinderMapPos, iType, item);
+	for (const auto& spectator : spectators) {
+		spectator->onRemoveTileItem(getTile(), thingMapPos, iType, item);
 	}
 
 	if (!hasFlag(TILESTATE_PROTECTIONZONE) || getBoolean(ConfigManager::CLEAN_PROTECTION_ZONES)) {
 		auto items = getItemList();
 		if (!items || items->empty()) {
-			g_game.removeTileToClean(this);
+			g_game.removeTileToClean(getTile());
 			return;
 		}
 
 		bool ret = false;
-		for (auto toCheck : *items) {
+		for (const auto& toCheck : *items) {
 			if (toCheck->isCleanable()) {
 				ret = true;
 				break;
 			}
 		}
 
 		if (!ret) {
-			g_game.removeTileToClean(this);
+			g_game.removeTileToClean(getTile());
 		}
 	}
 }
 
 void Tile::onUpdateTile(const SpectatorVec& spectators)
 {
-	const Position& cylinderMapPos = getPosition();
+	const Position& thingMapPos = getPosition();
 
 	// send to clients
-	for (Creature* spectator : spectators) {
-		assert(dynamic_cast<Player*>(spectator) != nullptr);
-		static_cast<Player*>(spectator)->sendUpdateTile(this, cylinderMapPos);
+	for (const auto& spectator : spectators) {
+		assert(spectator->getPlayer() != nullptr);
+		std::static_pointer_cast<Player>(spectator)->sendUpdateTile(getTile(), thingMapPos);
 	}
 }
 
-ReturnValue Tile::queryAdd(int32_t, const Thing& thing, uint32_t, uint32_t flags, Creature*) const
+ReturnValue Tile::queryAdd(int32_t, std::shared_ptr<const Thing> thing, uint32_t, uint32_t flags,
+                           std::shared_ptr<Creature>) const
 {
-	if (const Creature* creature = thing.getCreature()) {
+	if (auto creature = thing->getCreature()) {
```

Comment:
Could this be `const auto`
```suggestion
	if (const auto creature = thing->getCreature()) {
```